oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/resourceserver/authentication/JwtEncodedOAuth2AccessTokenAuthenticationProviderTests.java

```diff
@@ -0,0 +1,348 @@
+		}
+
+		@Bean
+		protected AuthenticationProvider oauth2AuthenticationProvider()
+		{
+			return new JwtEncodedOAuth2AccessTokenAuthenticationProvider(this.jwtDecoder());
+		}
+
+		@Bean
+		JwtDecoder jwtDecoder() {
+			HttpUrl url = this.server.url("/.well-known/jwks.json");
+
+			return new NimbusJwtDecoderJwkSupport(
+					url.toString(),
+					JWSAlgorithm.RS256.getName());
+		}
+
+	}
+
+	@Test
+	public void performWhenBearerIsSignedWithJwkOverEcdsaThenAuthorized()
+			throws Exception {
+
+		this.register(WebServerConfig.class, EcConfig.class);
+
+		JwkSetBuilder good = new JwkSetBuilder().ec("good");
+
+		String authority = new JwsBuilder(JWSAlgorithm.ES512)
+				.expiresAt(Instant.now().plus(1, ChronoUnit.HOURS))
+				.scope("permission.read")
+				.sign(good).serialize();
+
+		this.jwks(good);
+
+		this.mvc.perform(bearer(get("/"), authority))
+				.andExpect(content().string("OK"));
+
+		assertThat(this.server.getRequestCount()).isEqualTo(1);
+	}
+
+	@EnableWebSecurity
+	public static class EcConfig extends DefaultResourceServerConfig {
+		@Autowired MockWebServer server;
+
+		@Override
+		protected void configure(HttpSecurity http) throws Exception {
+			super.configure(http);
+
+			http
+				.authorizeRequests().anyRequest()
+					.access("authentication.hasClaim('scope', 'permission.read')");
+		}
+
+		@Bean
+		protected AuthenticationProvider oauth2AuthenticationProvider()
+		{
+			return new JwtEncodedOAuth2AccessTokenAuthenticationProvider(this.jwtDecoder());
+		}
+
+		@Bean
+		JwtDecoder jwtDecoder() {
+			HttpUrl url = this.server.url("/.well-known/jwks.json");
+
+			return new NimbusJwtDecoderJwkSupport(
+					url.toString(),
+					JWSAlgorithm.ES512.getName());
+		}
+
+	}
+
+	@Configuration
+	public static class WebServerConfig {
+		@RestController
+		public class SecuredController {
+			@GetMapping("/")
+			public String ok() {
+				return "OK";
+			}
+		}
+
+		@Bean
+		MockWebServer server() {
+			return new MockWebServer();
+		}
+	}
+
+	private MockHttpServletRequestBuilder bearer(MockHttpServletRequestBuilder mock, String authority) {
+		return mock.header("Authorization", "Bearer " + authority);
+	}
+
+	private void jwks(JwkSetBuilder builder) {
+		MockResponse response = new MockResponse()
+				.setHeader(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON_VALUE)
+				.setBody(builder.toJsonString());
+
+		this.server.enqueue(response);
+		this.server.enqueue(response);
+
+	}
+
+	private void register(Class<?>... classes) {
+		this.context = new AnnotationConfigWebApplicationContext();
+		this.context.register(classes);
+		this.context.setServletContext(new MockServletContext());
+		this.context.setServletConfig(new MockServletConfig());
+		this.context.refresh();
+		this.mvc = MockMvcBuilders.webAppContextSetup(this.context)
+								.apply(springSecurity()).build();
+
+		this.server = this.context.getBean(MockWebServer.class);
+	}
+
+	@Mock
+	JwtDecoder jwtDecoder;
```

Comment:
Member's are typically defined at the top of the class.

Reply:
Oh, right, this is something I was going to ask you about. So, half of these tests are traditional unit tests and the other half are integration tests. Because of this, they are mutually exclusive. 

What are we typically doing in this case? I can separate them into two different classes, though that doesn't seem to be the pattern I've run into so far in the code.

Note, my background is more along the lines that we have xxxTest for unit tests and xxxIT for integration tests, so that's the kind of lens that I'm looking through.

Reply:
Our integration tests typically use `MockMvc` and ensure that the `WebSecurityConfigurerAdapter` is configured and behaving as expected.

For Resource Server, we will eventually have `ResourceServerSecurityConfigurer` that is applied to the `HttpSecurity` builder. The integration (or security configuration) tests will load a specific `WebSecurityConfigurerAdapter` that has applied a specific configuration via `ResourceServerSecurityConfigurer` or `http.oauth2().resourceServer()` and conduct it's test.

Since we don't have a `ResourceServerSecurityConfigurer` as of yet, I would extract those `MockMvc` tests and put them in a class name called `ResourceServerSecurityConfigurerTests`. You can see an example of that in [OAuth2ClientConfigurerTests](https://github.com/spring-projects/spring-security/pull/5084/files#diff-b8c1b35ce9ce8d4fd3538da6e45e889bR65)